source/common/config/incremental_subscription_impl.h

```diff
@@ -0,0 +1,223 @@
+ */
+template <class ResourceType>
+class IncrementalSubscriptionImpl
+    : public Subscription<ResourceType>,
+      public GrpcStream<envoy::api::v2::IncrementalDiscoveryRequest,
+                        envoy::api::v2::IncrementalDiscoveryResponse, ResourceNameDiff> {
+public:
+  IncrementalSubscriptionImpl(const LocalInfo::LocalInfo& local_info,
+                              Grpc::AsyncClientPtr async_client, Event::Dispatcher& dispatcher,
+                              const Protobuf::MethodDescriptor& service_method,
+                              Runtime::RandomGenerator& random, Stats::Scope& scope,
+                              const RateLimitSettings& rate_limit_settings, SubscriptionStats stats)
+      : GrpcStream<envoy::api::v2::IncrementalDiscoveryRequest,
+                   envoy::api::v2::IncrementalDiscoveryResponse, ResourceNameDiff>(
+            std::move(async_client), service_method, random, dispatcher, scope,
+            rate_limit_settings),
+        type_url_(Grpc::Common::typeUrl(ResourceType().GetDescriptor()->full_name())),
+        local_info_(local_info), stats_(stats) {
+    request_.set_type_url(type_url_);
+    request_.mutable_node()->MergeFrom(local_info_.node());
+  }
+
+  // Enqueues and attempts to send a discovery request, (un)subscribing to resources missing from /
+  // added to the passed 'resources' argument, relative to resources_. Updates resources_ to
+  // 'resources'.
+  void buildAndQueueDiscoveryRequest(const std::vector<std::string>& resources) {
+    ResourceNameDiff diff;
+    std::set_difference(resources.begin(), resources.end(), resource_names_.begin(),
+                        resource_names_.end(), diff.added_.begin());
+    std::set_difference(resource_names_.begin(), resource_names_.end(), resources.begin(),
+                        resources.end(), diff.removed_.begin());
+
+    for (const auto& added : diff.added_) {
+      resources_[added] = "0";
+      resource_names_.insert(added);
+    }
+    for (const auto& removed : diff.removed_) {
+      resources_.erase(removed);
+      resource_names_.erase(removed);
+    }
+    queueDiscoveryRequest(diff);
+  }
+
+  void sendDiscoveryRequest(const ResourceNameDiff& diff) override {
+    if (!grpcStreamAvailable()) {
+      // Don't immediately try to reconnect: we rely on retry_timer_ for that.
+      ENVOY_LOG(debug, "No stream available to sendDiscoveryRequest for {}", type_url_);
+      return;
+    }
+    if (paused_) {
+      ENVOY_LOG(trace, "API {} paused during sendDiscoveryRequest().", type_url_);
+      return;
+    }
+
+    request_.clear_resource_names_subscribe();
+    request_.clear_resource_names_unsubscribe();
+    std::copy(diff.added_.begin(), diff.added_.end(),
+              request_.mutable_resource_names_subscribe()->begin());
+    std::copy(diff.removed_.begin(), diff.removed_.end(),
+              request_.mutable_resource_names_unsubscribe()->begin());
```

Comment:
Similarly to the above, I had to use ```RepeatedFieldBackInserter``` to avoid segfaults (for both resource_names_subscribe and resource_names_unsubscribe):
```
std::copy(diff.removed_.begin(), diff.removed_.end(),
  ::google::protobuf::RepeatedFieldBackInserter(request_.mutable_resource_names_unsubscribe()));
```

Reply:
Done.